drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,338 @@
+ * @file
+ * @{
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "hmc5883l_regs.h"
+#include "hmc5883l.h"
+
+#include "log.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+
+#define ASSERT_PARAM(cond) \
+    if (!(cond)) { \
+        DEBUG("[hmc5883l] %s: %s\n", \
+              __func__, "parameter condition (" # cond ") not fulfilled"); \
+        assert(cond); \
+    }
+
+#define DEBUG_DEV(f, d, ...) \
+        DEBUG("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+              __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#else /* ENABLE_DEBUG */
+
+#define ASSERT_PARAM(cond) assert(cond);
+#define DEBUG_DEV(f, d, ...)
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#define EXEC_RET(f, r) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return res; \
+    }
+
+#define EXEC_RET_CODE(f, r, c) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return c; \
+    }
+
+/** Forward declaration of functions for internal use */
+
+static int _is_available(const hmc5883l_t *dev);
+
+static uint8_t _get_reg_bit(uint8_t byte, uint8_t mask);
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit);
+static int _reg_read(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _reg_write(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+
+int hmc5883l_init(hmc5883l_t *dev, const hmc5883l_params_t *params)
+{
+    int res = HMC5883L_OK;
+
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(params != NULL);
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init sensor data structure */
+    dev->params = *params;
+
+    /* check availability of the sensor */
+    EXEC_RET(_is_available(dev), res)
+
+    uint8_t cfg_a = 0;
+    uint8_t cfg_b = 0;
+    uint8_t mode = 0;
+
+    /* set configuration register A and B */
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MA, params->meas_avg);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MS, params->meas_mode);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_DO, params->dor);
+    _set_reg_bit(&cfg_b, HMC5883L_REG_CFG_B_GN, params->gain);
+
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_A, &cfg_a, 1), res);
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_B, &cfg_b, 1), res);
+
+    /* set operation mode */
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_HS, 0);
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_MD, params->op_mode);
+
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_MODE, &mode, 1), res);
+
+    /* wait 6 ms accoring to data sheet */
+    xtimer_usleep(6 * US_PER_MS);
+
+    return res;
+}
+
+int hmc5883l_data_ready(const hmc5883l_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
```

Comment:
personally I prefer a simple `assert(dev != NULL)`. IMO the added debug output message is only taking ROM for limited additional benefit, as the failed assertion on its own is already pretty obvious what has happened...

Reply:
OK, my heart does not hang on the additional error message. The developer can take a look into the source code when the assertion fails.

Reply:
Exactly - and the developer has to do that in any case when debugging :-)